### PR TITLE
Add version pinning functionality

### LINKED ISSUE #72: Skipping casks when upgrading all
In cask there is option `brew pin` wich prevents `brew upgrade` from updating some cask.
(more info here https://github.com/Homebrew/legacy-homebrew/issues/35714)

I manually pinned cask via `brew pin caskroom/cask/jing` but `brew cu` still trying to update it!

Maybe skip pinned casks?

### LINKED ISSUE #90: Improvement: Blacklist not-to-be-updated software
It would be nice if one could set a certain software to be ignore with the updates;

I usually like to clean my software all up to date, but some software just can't be updated (plugins that depend on older versions, licensing reasons, etc.

### LINKED ISSUE #93: Exclude specific casks
Hey, is it possible to add an option to exclude certain casks from upgrading?

Something like `--exclude [casknames]`

The microsoft-office cask upgrade always hangs on the installer and I can't find a way to fix ist. Therefore it would be nice to simply exclude it.

### LINKED ISSUE #108: ignore some app to prevent check for update
Is it possible i ignore some application to prevent update ?

Thanks .

### PR DESCRIPTION
**Pull request by dongyuzheng**
# Why

I found myself needing this when I didn't want to update `java8`.

This PR resolves #108, resolves #93, resolves #90, resolves #72

# Not included in this PR

Currently you can pin casks that are not installed. So you can pin `asdasfsadasd`

# Usage

```
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --help
...

OPTIONS:
...
    If --pinned is passed, print all pinned apps

    If --pin CASK is passed, pin the current app version

    If --unpin CASK is passed, unpin the current app version
```

# Example usage

```
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --pinned
java8
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --pin firefox
Pinned: firefox
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --pin firefox
Already pinned: firefox
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --pinned
java8
firefox
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --unpin firefox
Unpinned: firefox
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --unpin firefox
Not pinned: firefox
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --pinned
java8
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu
==> Options
Include auto-update (-a): false
Include latest (-f): false
==> Updating Homebrew
Already up-to-date.
==> Finding outdated apps
       Cask                    Current                  Latest                A/U    Result
...
 8/19  java8                   1.8.0_192-b12,750e1c...  8.202.03,07_nov_2018       [ PINNED ]
...
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu --unpin java8
Unpinned: java8
gary@v1020-wn-17-117:~/pg/homebrew-cask-upgrade (master)$ brew cu
==> Options
Include auto-update (-a): false
Include latest (-f): false
==> Updating Homebrew
Already up-to-date.
==> Finding outdated apps
       Cask                    Current                  Latest                A/U    Result
...
 8/19  java8                   1.8.0_192-b12,750e1c...  8.202.03,07_nov_2018       [OUTDATED]
...
==> Found outdated apps
     Cask   Current                                         Latest                A/U    Result
1/1  java8  1.8.0_192-b12,750e1c8617c5452694857ad95c3ee230  8.202.03,07_nov_2018       [OUTDATED]

Do you want to upgrade 1 app [y/N]? ^C
```

### CODE REVIEW

File: lib/bcu.rb [hunk 1]
```diff
@@ -82,8 +130,9 @@ def self.find_outdated_apps(quiet)
 
     installed.each do |app|
       version_latest = (app[:version] == "latest")
-
-      if options.force && options.all && version_latest && app[:auto_updates]
+      if pins.include?(app[:token])
+        state_info[app] = "pinned"
```
Comment (ondrejfuhrer) [inline lib/bcu.rb]: For some reason doesn't work properly under certain conditions.
I pinned 3 casks as following:
```bash
> brew cu --pins                                                                                                                    
toggl
gimp
fork
```

And when I run `brew cu --all --cleanup` this is what I get:
```bash
> brew cu --all --cleanup                                                                                                           
==> Options
Include auto-update (-a): true
Include latest (-f): false
==> Updating Homebrew
Already up-to-date.
==> Finding outdated apps
       Cask     Current     Latest     A/U    Result
...
21/62  fork     1.0.72.2    1.0.72.2    Y   [   OK   ]
22/62  gimp     2.10.8,2    2.10.8,2        [ PINNED ]
...
53/62  toggl    7.4.330     7.4.331         [ PINNED ]
...
``` 

My guess is that the `auto-update` flag is somehow interfering with the pinning implementation.

**Comment by ondrejfuhrer** [comment]
Hey @dongyuzheng , thanks a lot for your contribution! It's nice that someone found a time to do that. Please do me a favour and change the description here so that instead of lust listing those issues change it to `Resolves: #...` so that when the PR is merged, those issues are automatically closed.

Thanks! 👍

**Comment by dongyuzheng** [comment]
Hi @ondrejfuhrer, does my latest commit give desired output for your case?

File: lib/bcu.rb (hunk 1 above)
Comment (ondrejfuhrer) [reply lib/bcu.rb]: Looks good now 👍

File: lib/bcu/options.rb [hunk 2]
```diff
@@ -45,6 +48,18 @@ def self.parse!(args)
       opts.on("--no-quarantine", "Add --no-quarantine option to install command, see brew cask documentation for additional information") do
         options.install_options += " --no-quarantine"
       end
+
+      opts.on("--pinned", "List pinned apps") do
+        options.list_pins = true
+      end
+
+      opts.on("--pin APP", "App to pin") do |app|
```
Comment (ondrejfuhrer) [inline lib/bcu/options.rb]: Everywhere we / Homebrew uses `CASK`, I don't see a reason why we should introduce new naming? I saw that in one of the versions you had used that, any particular reason why you decided to use different naming?

File: lib/bcu.rb [hunk 3]
```diff
@@ -4,11 +4,58 @@
...
+
+PINS_FILE = File.expand_path(File.dirname(__FILE__) + "/../pins")
 
 module Bcu
   def self.process(args)
     parse!(args)
 
+    FileUtils.touch(PINS_FILE)
+
+    pins = Set[]
```
Comment (ondrejfuhrer) [inline lib/bcu.rb]: Would be nice to extract this to a separate function that returns that set
```suggestion
    pins = load_pins
```

File: lib/bcu.rb [hunk 4]
```diff
@@ -4,11 +4,58 @@
...
+          f.puts(options.pin)
+        end
+        puts "Pinned: #{options.pin}"
+      end
+      return
+    end
+
+    if options.unpin
+      if pins.include?(options.unpin)
+        pins.delete(options.unpin)
```
Comment (ondrejfuhrer) [inline lib/bcu.rb]: Would be nice to extract that to a separate function that takes care of that.
```suggestion
        remove_pin options.unpin
```

The function is already quite big and if we could extract something out to increase readability, would be great. What do you think 🙂

File: README.md [hunk 5]
```diff
@@ -54,6 +54,9 @@ Usage: brew cu [CASK] [options]
     -y, --yes             Update all outdated apps; answer yes to updating packages.
     -q, --quiet           Do not show information about installed apps or current options.
         --no-quarantine   Pass --no-quarantine option to `brew cask install`.
+        --pinned          Print all pinned apps
+        --pin APP         Pin the current app version
```
Comment (ondrejfuhrer) [inline README.md]: Would be also great if you could write a short description into the README about the pinning, what does it do, etc. Also I would include a warning, that pinning only works if used with `brew cu`, if someone would use directly a `brew cask upgrade` command it won't have any effect.

File: lib/bcu/options.rb (hunk 2 above)
Comment (dongyuzheng) [reply lib/bcu/options.rb]: I saw that the code/documentation used `app` in some places, so I change them to `app` to be consistent. But I changed it to `cask` now.

**Comment by muescha** [comment]
i think the pinned file should be better in the `~/` folder. for example as `~/.brew-cu-pinned`

**Comment by ondrejfuhrer** [comment]
@muescha Not sure if that would make any difference. Not even the Homebrew core is using a home folder for that and they are pinning it within Homebrew structure, so I don't see any important reason why would you do that. 

Also if you would remove the tap it would leave those files on the home folder which is not very clean.

**Comment by yurikoles** [comment]
I think @ondrejfuhrer is right there is `--appdir` option, but it still uses global `Caskroom` for apps.

**Comment by muescha** [comment]
ok - we should stick with the homebrew core behaviour

**Pull request merged**